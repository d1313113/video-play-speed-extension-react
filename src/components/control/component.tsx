import React, { FC } from "react";
import { Slider, Button } from "antd";
import {
  PlayCircleOutlined,
  PauseCircleOutlined,
  AudioMutedOutlined,
  AudioOutlined
} from "@ant-design/icons";
import { VideoStatus } from "@/content-script/actions/videosPlaybackRate";
import "./styles.scss";

const RATES = [1, 2, 3, 4, 5];

export interface ControlProps {
  videoList: VideoStatus[];
  onSetRate: (idx: number, value: number) => Promise<void>;
  onPlay: (idx: number) => Promise<void>;
  onPause: (idx: number) => Promise<void>;
  onMuted: (idx: number, muted: boolean) => Promise<void>;
}

export const Control: FC<ControlProps> = (props: ControlProps) => {
  const handleSetVideoRate = (index: number, value: number): void => {
    props.onSetRate(index, value);
  };

  const handleClickRate = (index: number, rate: number): void => {
    props.onSetRate(index, rate);
  };

  const { videoList } = props;

  return (
    <main className="control">
      <div className="title">可拖动step调整速度</div>
      <section className="control-main">
        {videoList.map((item, index) => (
          <div className="video-item" key={index}>
            <Slider
              onChange={(value: number): void =>
                handleSetVideoRate(index, value)
              }
              min={0.1}
              max={5}
              step={0.1}
              defaultValue={1}
              value={item.playbackRate}
            />
            <div className="rate-wrapper">
              {item.paused ? (
                <PlayCircleOutlined
                  className="icon-control"
                  onClick={(): Promise<void> => props.onPlay(index)}
                />
              ) : (
                <PauseCircleOutlined
                  className="icon-control"
                  onClick={(): Promise<void> => props.onPause(index)}
                />
              )}
              {item.muted ? (
                <AudioMutedOutlined
                  className="icon-control middle"
                  onClick={(): Promise<void> =>
                    props.onMuted(index, !item.muted)
                  }
                />
              ) : (
                <AudioOutlined
                  className="icon-control middle"
                  onClick={(): Promise<void> =>
                    props.onMuted(index, !item.muted)
                  }
                />
              )}
              {RATES.map(rate => (
                <Button
                  key={rate}
                  onClick={(): void => handleClickRate(index, rate)}
                  size="small"
                  type={rate === 1 ? "primary" : "default"}
                  disabled={rate === item.playbackRate}
                >
                  {rate === 1 ? "Reset" : `✖️${rate}`}
                </Button>
              ))}
            </div>
          </div>
        ))}
      </section>
    </main>
  );
};
