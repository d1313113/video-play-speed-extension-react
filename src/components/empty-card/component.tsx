import React, { FC } from "react";
import { Empty } from "antd";
import "./styles.scss";

export const EmptyCard: FC = () => (
  <section className="empty-wrapper">
    <Empty className="empty" description="当前页面未找到视频:)" />
  </section>
);
