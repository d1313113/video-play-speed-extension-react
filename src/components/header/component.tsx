import React, { FC, useState } from "react";
import { ReloadOutlined } from "@ant-design/icons";
import "./styles.scss";

interface HeaderProps {
  onReload: () => void;
}

export const Header: FC<HeaderProps> = (props: HeaderProps) => {
  const [spin, setSpin] = useState<boolean>(false);

  const handleReload = (): void => {
    setSpin(!spin);
    if (!spin) {
      props.onReload();
      setTimeout((): void => {
        setSpin(false);
      }, 1000);
    }
  };

  return (
    <header className="header-wrapper">
      <div className="header-content">
        <div className="title">视频播放速度调节插件</div>
        <div className="sub-title">可以让你轻松调整视频播放速度</div>
      </div>
      <ReloadOutlined
        onClick={handleReload}
        className="icon-reload"
        spin={spin}
        style={{
          fontSize: "20px"
        }}
      />
    </header>
  );
};
