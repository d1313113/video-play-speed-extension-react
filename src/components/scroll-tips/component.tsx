import React, { FC } from "react";
import { DoubleRightOutlined } from "@ant-design/icons";
import "./styles.scss";

export const ScrollTips: FC = () => (
  <section className="tips-wrapper">
    <div className="tips">Scroll to load more</div>
    <DoubleRightOutlined
      className="icon-tips"
      rotate={90}
      style={{ fontSize: "20px" }}
    />
  </section>
);
