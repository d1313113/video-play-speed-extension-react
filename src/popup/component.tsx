import React, { FC, useEffect, useState } from "react";
import { browser } from "webextension-polyfill-ts";
import { throttle } from "lodash";
import getCurrentTab from "@/utils/getCurrentTab";
import {
  actionTypes,
  VideoStatus,
  $$
} from "@/content-script/actions/videosPlaybackRate";

import { Header } from "@/components/header";
import { EmptyCard } from "@/components/empty-card";
import { Control } from "@/components/control";
import { ScrollTips } from "@/components/scroll-tips";
import "./styles.scss";

export const Popup: FC = () => {
  const [videoList, setVideoList] = useState<VideoStatus[]>([]);
  const [scrollTop, setScrollTop] = useState<number>(0);

  const initPopUp = async (): Promise<void> => {
    console.log(11111);
    const { id: tabId } = await getCurrentTab();
    if (tabId != null) {
      const resp: VideoStatus[] = await browser.tabs.sendMessage(tabId, {
        action: actionTypes.GET_VIDEOS_STATUS
      });
      setVideoList(resp);
    }
  };

  const handleSetVideoRate = async (
    idx: number,
    value: number
  ): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PLAYBACK_RATE,
        idx,
        value
      });
      initPopUp();
    }
  };

  const handleSetVideoPlay = async (idx: number): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PLAY,
        idx
      });
      initPopUp();
    }
  };

  const handleSetVideoPause = async (idx: number): Promise<void> => {
    const { id: tabId } = await getCurrentTab();

    if (tabId != null) {
      await browser.tabs.sendMessage(tabId, {
        action: actionTypes.SET_VIDEO_PAUSE,
        idx
      });
      initPopUp();
    }
  };

  const handleScroll = (html: HTMLElement): void => {
    try {
      const { scrollTop } = html;
      setScrollTop(scrollTop);
    } catch (e) {
      console.error(e);
    }
  };

  // Sends the `popupMounted` event
  useEffect(() => {
    browser.runtime.sendMessage({ popupMounted: true });
    initPopUp();
    const html = $$<HTMLElement>("html")[0];
    const _throttle = throttle((): void => handleScroll(html), 100);
    window.addEventListener("scroll", _throttle, false);

    return (): void => {
      window.removeEventListener("scroll", _throttle, false);
    };
  }, []);

  // Renders the component tree
  return (
    <div className="popup-container">
      <div className="container">
        <Header onReload={initPopUp} />
        {!videoList.length ? (
          <EmptyCard />
        ) : (
          <Control
            videoList={videoList}
            onSetRate={handleSetVideoRate}
            onPlay={handleSetVideoPlay}
            onPause={handleSetVideoPause}
          />
        )}
        {!!videoList.length && videoList.length > 5 && scrollTop < 30 && (
          <ScrollTips />
        )}
      </div>
    </div>
  );
};
