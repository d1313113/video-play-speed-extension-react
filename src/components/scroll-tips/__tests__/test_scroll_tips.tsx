import * as React from "react";
import { ScrollTips } from "../component";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";

it("component renders", () => {
  const tree:
    | ReactTestRendererJSON
    | ReactTestRendererJSON[]
    | null = renderer.create(<ScrollTips />).toJSON();
  expect(tree).toMatchSnapshot();
});
