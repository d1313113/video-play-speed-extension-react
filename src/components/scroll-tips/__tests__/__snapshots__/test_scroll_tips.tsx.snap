// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
<section
  className="tips-wrapper"
>
  <div
    className="tips"
  >
    Scroll to load more
  </div>
  <span
    aria-label="double-right"
    className="anticon anticon-double-right icon-tips"
    role="img"
    style={
      Object {
        "fontSize": "20px",
      }
    }
  >
    <svg
      aria-hidden="true"
      className=""
      data-icon="double-right"
      fill="currentColor"
      focusable="false"
      height="1em"
      style={
        Object {
          "msTransform": "rotate(90deg)",
          "transform": "rotate(90deg)",
        }
      }
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M533.2 492.3L277.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H188c-6.7 0-10.4 7.7-6.3 12.9L447.1 512 181.7 851.1A7.98 7.98 0 00188 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5zm304 0L581.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H492c-6.7 0-10.4 7.7-6.3 12.9L751.1 512 485.7 851.1A7.98 7.98 0 00492 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5z"
      />
    </svg>
  </span>
</section>
`;
